docs/basics.rst

```diff
@@ -411,3 +411,24 @@ production environments for reproducible builds.
     If you'd like a ``requirements.txt`` output of the lockfile, run ``$ pipenv lock -r``.
     This will not include hashes, however. To get a ``requirements.txt``
     you can also use ``$ pipenv run pip freeze``.
+
+
+☤ Pipenv and Docker Conatiners
+------------------------------
+
+In general, you should not need to install pipenv inside docker container to manage
+the dependencies of your software. Instead you can generate a ``requirements.txt`` as described
+above and use it inside your container, to install the dependencies, e.g.:
+
+  $ pipenv lock -r > requirements.txt
```

Comment:
@oz123 these instructions assume that you've got `pipenv` installed outside of docker? That doesn't feel desirable to me.

Reply:
is meant for managing virtual environment*s*. Usually, a container contains just one. That is the reason why you probably don't need it inside the container.
If you really desire you can use multi stage build with one stage building your vent with pipenv and then copying it to the end container. But that's way out of scope of the manual and I assumed that power users of docker aren't the audience of these instructions.
That said, what is your suggestion? Have I overlooked something? Please let me know if I can improve the documentation.

Reply:
What about having `pipenv` installed in the docker container and running `pipenv sync`?   Seems about half of the time docker containers are using virtualenvs and half the time they aren't.  At my last job we buil ECR images in docker, and so having to run a command outside of docker in the CI to generate the latest requirements.txt from outside of the container, well we wouldn't have done it that way since the whole thing ran the build in docker.

Reply:
I still don't get why one would want to have ~80mb of pipenv inside the container image.
Even if you run your pipeline in container one step can generate requirements.txt and the next one can pickup the file and install the dependencies.
And these steps can be in Jenkins/giltab/GitHub/your-ci or as a I mentioned multi stage docker file. 
As side note: there is some confusion always regarding the terminology.
The docker file is used to build the image .
The container is spawned from it. 
You install pipenv during the image build. And then you have it in all the containers built from it.
Pipenv is a build tool, like npm or make or GCC. IMHO it does not belong in the final container.
Whether you exclude it with my first suggestion or later is eventually the same. I chose to document the most easy one. 
I will add a note about pipenv exclusion from the final container.

Reply:
Maybe I am missing something @oz123 but pipenv wheel is < 4mb, so having it in the final container or not seems like a matter of preference.  I was just stating that in my prior case, we were using a local user and virtualenvs within the final container, and while we weren't using `pipenv` at all, if we were, we likely would have converted from `requirements.txt` to `Pipfile` and `Pipfile.lock` and generate the virtualenv using `pipenv sync` to build the dependencies in the lock file during CI build (the Docker files building the image).  There is no need to generate a requirements.txt at all the way I see it, if you are relying on your lock file.  

Also and **most importantly** I just tried the example of `pipenv lock -r > requirements.txt` and it doesn't include any hash values in the requirements.txt so doing it this way basically negates any protection of having the lock file and the benefits of using `pipenv sync` during the build.

Reply:
The size I mentioned is wrong (based on the package in gentoo with compiled __pycache__ include). Sorry about that.
As for the hashing, yes, you are right, requirements.txt does not include the hashes.
Thanks for pointing this out. I will update the docs accordingly, and suggest to use multistage build which exclude pipenv from the final image.

Reply:
@oz123, I agree that multistage builds feel like The Right Way to do this. Hopefully it's not too much work to write up. Thanks for improving the documentation!

Reply:
I would also modify `pipenv lock -r` to add hashes:
see #4189.